056-zk-kraft-migration.md

```diff
@@ -0,0 +1,453 @@
+More specifically, the controllers are automatically updated with the following configuration (the user cannot set them manually).
+
+```shell
+# Disable the migration
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+```
+
+Finally, the operator rolls the controllers, via the `KafkaRoller`, and waits for them to be ready again.
+They are now disconnected from the ZooKeeper ensemble which remain running but are no longer the source of truth for Kafka metadata.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves to the `FullKRaft` state and ZooKeeper is not used anymore.
+
+#### Phase 5: Removing ZooKeeper
+
+At this point the ZooKeeper ensemble is out of the picture, still running but not used anymore.
+
+**_Preconditions_**
+
+The FSM is in the `FullKRaft` state (as per `Kafka.status.migrationState` field) and the operator expects the user to finalize the migration by removing ZooKeeper.
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: enabled` annotation and removes the `spec.zookeeper` on the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation and, via the `KafkaReconciler` component, deletes the ZooKeeper nodes pods.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to a `NoMigration` state (or null) and the operator upgrades `Kafka.status.migrationState` field accordingly.
+
+### Metrics
+
+During ZooKeeper to KRaft migration, there are some Kafka metrics that could be useful to check the status in each of the phases.
+More details in the [“Metrics”](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-Metrics) section of KIP-866.
+
+The most interesting and useful ones are:
+
+| MBean                                                             | Description                                                                                                                        |
+|-------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
+| kafka.controller:type=KafkaController,name=MetadataType           | An enumeration of: ZooKeeper (1), KRaft (2), or Dual (3). The active controller reports this.                                      |
+| kafka.controller:type=KafkaController,name=MigratingZkBrokerCount | A count of ZK brokers that are registered with KRaft and ready for migration. This will only be reported by the active controller. |
+| kafka.controller:type=KafkaController,name=ZkMigrationState       | An enumeration of the possible migration states the cluster can be in. This is only reported by the active controller.             |
```

Comment:
I think this is also an important metric to monitor when users decided to rollback to ZK mode. They need to make sure there's no lag between ZK and KRaft metadata, to avoid data lose after rollback.

`kafka.controller:type=KafkaController,name=ZkWriteBehindLag`. 

`The amount of lag in records that ZooKeeper is behind relative to the highest committed record in the metadata log. This metric will only be reported by the active KRaft controller.`

Reply:
This is a good point ...
Is it possible for this metric to be 0? Because I would assume that avoiding data loss means all metadata records in KRaft are written to ZooKeeper as well (while in "dual write" mode). It would assume that in order to rollback, the user should not use the cluster for administrative operations which impact on metadata otherwise this lag will keep being not 0. Is my assumption right?

At FSM level, my thought is that in all the states where rollback is possible, an additional precondition is that `ZkWriteBehindLag` is 0. If it's not, the user should be warned in the `Kafka` resource state and the operator would try again on the next reconciliation.

Reply:
> Is it possible for this metric to be 0? 

Yes, it will be 0 if no lag between ZK/KRaft.

>  It would assume that in order to rollback, the user should not use the cluster for administrative operations which impact on metadata otherwise this lag will keep being not 0. Is my assumption right?

Yes, that's the best practice to avoid potential data loss.

> At FSM level, my thought is that in all the states where rollback is possible, an additional precondition is that ZkWriteBehindLag is 0. If it's not, the user should be warned in the Kafka resource state and the operator would try again on the next reconciliation.

Yes, make sense.

Reply:
@showuon thinking more about this ... I was wondering if, instead of using this metric as a precondition, we just set a warning in the `Kafka` resource status when it's not 0. I mean, what if the user wants anyway rolls back accepting metadata loss?

Reply:
Fine with me, as long as we explain clearly in the proposal and log it.

Reply:
Done.